src/cryptonote_protocol/cryptonote_protocol_handler.inl

```diff
@@ -1565,6 +1691,37 @@ namespace cryptonote
               m_block_queue.remove_spans(span_connection_id, start_height);
               return 1;
             }
+            else if (bvc.m_added_to_main_chain)
+            {
+              block tmp;
+              const std::vector<crypto::hash>* tx_hashes_ptr = nullptr;
+              if (!pblocks.empty())
+              {
+                tx_hashes_ptr = &pblocks[blockidx].tx_hashes;
+              }
+              else
+              {
+                crypto::hash ignore;
+                if (parse_and_validate_block_from_blob(block_entry.block, tmp, &ignore))
+                  tx_hashes_ptr = &tmp.tx_hashes;
+              }
+              if (tx_hashes_ptr)
+              {
+                for (const auto &h : *tx_hashes_ptr)
+                  if (m_request_manager.already_requested_tx(h))
+                  {
+                    boost::uuids::uuid owner;
+                    if (m_request_manager.get_current_request_peer_id(h, owner) && !owner.is_nil())
+                    {
+                        m_p2p->for_connection(owner, [&](cryptonote_connection_context &ctx, nodetool::peerid_type, uint32_t)->bool {
+                            ctx.remove_in_flight_request();
```

Comment:
If you call `remove_in_flight_request` here, but that peer already received it and we haven't got the response yet (ping is not 0, so network delays are in play), will it `remove_in_flight_request` again when this response arrives?